src/common.ts

```diff
@@ -54,4 +55,58 @@ export class Common {
     logger.debug(`loginUrl: ${loginUrl}`);
     return loginUrl;
   }
+
+  public static async identifyPossibleScratchOrgs(fields: AuthFields, orgAuthInfo: AuthInfo): Promise<void> {
+    const logger = await Logger.child('Common', { tag: 'identifyPossibleScratchOrgs' });
+
+    // return if we already know the hub or we know it is a devhub or prod-like
+    if (fields.isDevHub || fields.devHubUsername || fields.loginUrl === 'https://login.salesforce.com') return;
+
+    // there are no hubs to ask, so quit early
+    if (!(await AuthInfo.hasAuthentications())) return;
+    logger.debug('getting devhubs from authfiles');
+
+    // TODO: return if url is not sandbox-like to avoid constantly asking about production orgs
+    // TODO: someday we make this easier by asking the org if it is a scratch org
+
+    const hubAuthInfos = await this.getDevHubAuthInfos();
+    logger.debug(`found ${hubAuthInfos.length} DevHubs`);
+    if (hubAuthInfos.length === 0) return;
+
+    // ask all those orgs if they know this orgId
+    await Promise.all(
+      hubAuthInfos.map(async (hubAuthInfo) => {
+        try {
+          const devHubOrg = await Org.create({ aliasOrUsername: hubAuthInfo.getUsername() });
+          const conn = devHubOrg.getConnection();
+          const data = await conn.query<QueryResult<{ Id: string }>>(
+            `select Id from ScratchOrgInfo where Id = '${orgAuthInfo.getFields().orgId}'`
+          );
+          if (data.totalSize > 0) {
+            // if any return a result
+            logger.debug(`found orgId ${fields.orgId} in devhub ${hubAuthInfo.getUsername()}`);
+            try {
+              await orgAuthInfo.save({ ...orgAuthInfo.getFields(), devHubUsername: hubAuthInfo.getUsername() });
+            } catch (error) {
+              logger.debug('error updating auth file');
+            }
+            logger.debug('updated authfile with devHubUsername');
+          }
+        } catch (error) {
+          logger.error(`Error connecting to query from ${hubAuthInfo.getUsername()}`);
+        }
+      })
+    );
+  }
+
+  public static async getDevHubAuthInfos(): Promise<AuthInfo[]> {
```

Comment:
@mshanemc how does this work? Looking at my local filesystem the files that have this pattern contain users associated with an org. Is the hope that the dev hub username will be present in one of these files?

Using this as an assumption I ran `find ~/.sfdx | grep 00D | xargs fgrep admin@integrationtesthubna40.org` locally but I did not find the username for a dev hub in any file that starts with `00D`.

Reply:
that regex filters out non-orgs (ie, user auth files).  I'm borrowing from here: https://github.com/salesforcecli/plugin-org/blob/9a2881bf6e6a84c1671cb9c3b36aadf7dc7ee93f/src/shared/orgListUtil.ts#L106

We have to actually do the AuthInfo.create and getFields to see if they self-identify as hubs.